### ISSUE TITLE
share() always returns a list

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5862]
Migrated from [rt.perl.org#16770](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=16770) (status was 'resolved')

Searchable as RT16770$

**Comment by p5pRT** [comment]
### From @schwern

#### Created by @schwern



The threads&#8203;::shared docs say&#8203;:

&nbsp;       share VARIABLE
&nbsp;           "share" takes a value and marks it as shared\. You can
&nbsp;           share a scalar\, array\, hash\, scalar ref\, array ref or
&nbsp;           hash ref\.  "share" will return the shared rvalue\.

but it always returns it as a list\.

\#\!/usr/bin/perl \-lw

use threads&#8203;::shared;

$scalar = share\(23\);
$hash   = share\(\{ foo => 23 \}\);
$array  = share\(\[1\,2\,3\]\);

print $scalar;  \# should be 23
print $hash;    \# should be a hash ref
print $array;   \# should be an array ref


$ bleadperl ~/tmp/bug\.plx
1
1
1

This makes idioms like this not work\.

&nbsp;     $foo\[42\] = &share\(\{ foo => 42\, bar => 23 \}\);


If the program is changed to this&#8203;:

\#\!/usr/bin/perl \-lw

use threads&#8203;::shared;

\($scalar\) = share\(23\);
\($hash\)   = share\(\{ foo => 23 \}\);
\($array\)  = share\(\[1\,2\,3\]\);

print $scalar;  \# should be 23
print $hash;    \# should be a hash ref
print $array;   \# should be an array ref

it works as expected\.

$ bleadperl ~/tmp/bug\.plx
23
HASH\(0x101d517c\)
ARRAY\(0x101d5284\)

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.9.0:

Configured by schwern at Mon Aug 26 00:43:26 PDT 2002.

Summary of my perl5 (revision 5.0 version 9 subversion 0 patch 17702) configuration:
  Platform:
    osname=linux, osvers=2.4.18, archname=ppc-linux-thread-multi-64int
    uname='linux blackrider 2.4.18 #1 thu may 16 00:32:38 edt 2002 ppc unknown unknown gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20011002 (Debian prerelease)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=87654321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL17702


@INC for perl v5.9.0:
    /usr/local/bleadperl/lib/5.9.0/ppc-linux-thread-multi-64int
    /usr/local/bleadperl/lib/5.9.0
    /usr/local/bleadperl/lib/site_perl/5.9.0/ppc-linux-thread-multi-64int
    /usr/local/bleadperl/lib/site_perl/5.9.0
    /usr/local/bleadperl/lib/site_perl
    .


Environment for perl v5.9.0:
    HOME=/home/schwern
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/usr/local/X11R6/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/home/schwern/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @ysth

On Wed\, 28 Aug 2002 06&#8203;:04&#8203;:18 \-0700\, schwern@&#8203;pobox\.com wrote&#8203;:
>This patch fixes threads&#8203;::shared's behavior when threads are not enabled and
>tests the behavior\.

>\-sub share\_disabled \{ return @&#8203;\_\}
>\+sub share\_disabled          \(\\\[$@&#8203;%\]\) \{ return $\_\[0\] \}

To have code that works both with and without this fix \(i\.e\. both 5\.8\.0 and
5\.8\.1\) I think you can say&#8203;:

&nbsp;    $Test\_Results\[$Curr\_Test\-1\] = $\{\\&share\(\{
&nbsp;        'ok'      => 1\,
&nbsp;        actual\_ok => 1\,
&nbsp;        name      => ''\,
&nbsp;        type      => 'skip'\,
&nbsp;        reason    => $why\,
&nbsp;    \}\)\};

**Comment by p5pRT** [comment]
### From @schwern

On Sun\, Sep 01\, 2002 at 09&#8203;:04&#8203;:27PM \-0700\, Yitzchak Scott\-Thoennes wrote&#8203;:
> On Wed\, 28 Aug 2002 06&#8203;:04&#8203;:18 \-0700\, schwern@&#8203;pobox\.com wrote&#8203;:
> >This patch fixes threads&#8203;::shared's behavior when threads are not enabled and
> >tests the behavior\.
> 
> >\-sub share\_disabled \{ return @&#8203;\_\}
> >\+sub share\_disabled          \(\\\[$@&#8203;%\]\) \{ return $\_\[0\] \}
> 
> To have code that works both with and without this fix \(i\.e\. both 5\.8\.0 and
> 5\.8\.1\) I think you can say&#8203;:
> 
>     $Test\_Results\[$Curr\_Test\-1\] = $\{\\&share\(\{
>         'ok'      => 1\,
>         actual\_ok => 1\,
>         name      => ''\,
>         type      => 'skip'\,
>         reason    => $why\,
>     \}\)\};

Yep\, that appears to work and it's shorter than my existing work\-arounds\.


\-\- 

Michael G\. Schwern   \<schwern@&#8203;pobox\.com>    http&#8203;://www\.pobox\.com/~schwern/
Perl Quality Assurance      \<perl\-qa@&#8203;perl\.org>         Kwalitee Is Job One
It's Yellowing Laudanum time\!

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From arthur@contiller.se

Solved a long time ago by patch \#17810

Thanks schwern\.

Arthur

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'
